tempesta_fw/classifier/frang.c

```diff
@@ -1007,384 +975,27 @@ static TfwClassifier frang_class_ops = {
-	},
-	{
-		.name = "connection_rate",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.conn_rate,
-	},
-	{
-		.name = "connection_burst",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.conn_burst,
-	},
-	{
-		.name = "concurrent_connections",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.conn_max,
-	},
-	{
-		.name = "client_header_timeout",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = (unsigned int *)&frang_cfg.clnt_hdr_timeout,
-	},
-	{
-		.name = "client_body_timeout",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = (unsigned int *)&frang_cfg.clnt_body_timeout,
-	},
-	{
-		.name = "http_uri_len",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_uri_len,
-	},
-	{
-		.name = "http_field_len",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_field_len,
-	},
-	{
-		.name = "http_body_len",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_body_len,
-	},
-	{
-		.name = "http_header_cnt",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_hdr_cnt,
-	},
-	{
-		.name = "http_header_chunk_cnt",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_hchunk_cnt,
-	},
-	{
-		.name = "http_body_chunk_cnt",
-		.deflt = "0",
-		.handler = tfw_cfg_set_int,
-		.dest = &frang_cfg.http_bchunk_cnt,
-	},
-	{
-		.name = "http_host_required",
-		.deflt = "true",
-		.handler = tfw_cfg_set_bool,
-		.dest = &frang_cfg.http_host_required,
-	},
-	{
-		.name = "http_ct_required",
-		.deflt = "false",
-		.handler = tfw_cfg_set_bool,
-		.dest = &frang_cfg.http_ct_required,
-	},
-	{
-		.name = "http_methods",
-		.deflt = "",
-		.handler = frang_cfgop_http_methods,
-		.cleanup = frang_cfgop_cleanup_http_methods,
-	},
-	{
-		.name = "http_ct_vals",
-		.deflt = NULL,
-		.handler = frang_cfgop_http_ct_vals,
-		.allow_none = 1,
-		.cleanup = frang_cfgop_cleanup_http_ct_vals,
-	},
-	{
-		.name = "http_resp_code_block",
-		.deflt = NULL,
-		.handler = frang_cfgop_rsp_code_block,
-		.dest = &frang_cfg,
-		.allow_none = 1,
-		.cleanup = frang_cfgop_cleanup_rsp_code_block,
-	},
-	{ 0 }
-};
-
-static TfwCfgSpec frang_specs[] = {
-	{
-		.name = "frang_limits",
-		.handler = tfw_cfg_handle_children,
-		.cleanup = tfw_cfg_cleanup_children,
-		.dest = frang_limits_specs,
-	},
-	{ 0 }
-};
-
-static TfwMod frang_mod = {
-	.name	= "frang",
-	.start	= frang_start,
-	.specs	= frang_specs,
-};
-
 static int __init
 frang_init(void)
 {
 	int r;
 
 	BUILD_BUG_ON((sizeof(FrangAcc) > sizeof(TfwClassifierPrvt)));
 
-	tfw_mod_register(&frang_mod);
 	tfw_classifier_register(&frang_class_ops);
 
 	r = tfw_gfsm_register_fsm(TFW_FSM_FRANG_REQ, frang_http_req_handler);
 	if (r) {
-		TFW_ERR_NL("frang: can't register fsm\n");
+		TFW_ERR_NL("frang: can't register request fsm\n");
 		goto err_fsm;
 	}
 
+	r = tfw_gfsm_register_fsm(TFW_FSM_FRANG_RESP, frang_resp_handler);
```

Comment:
Previously the hook was added only if needed, from now on - it's loaded on start. As @krizhanovsky suggested to move
frang to `tempesta_fw` module, both hooks will be loaded on every start. But we can move the  gfsm callbacks registration to `cfgend()` and register them only if frang is configured.

Reply:
Here the matter is significantly complicated by the support for reconfiguration for local Frang lmits - we must separately track the registration/removal of GFSM callbacks for the current and new configurations, and separately track for global non-configurable Framg limits, and besides, the GFSM functionality must be adapted for FSMs registration/removal on the fly.